exonum_derive/src/lib.rs

```diff
@@ -0,0 +1,302 @@
+// Copyright 2018 The Exonum Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![recursion_limit = "256"]
+
+extern crate proc_macro;
+extern crate proc_macro2;
+
+extern crate syn;
+#[macro_use]
+extern crate quote;
+
+use proc_macro::TokenStream;
+use proc_macro2::{Ident, Span};
+use syn::{Data, DeriveInput, Lit, Meta, NestedMeta, Path};
+
+use std::env;
+
+#[proc_macro_derive(ProtobufConvert, attributes(protobuf_convert))]
+pub fn protobuf_convert_derive(input: TokenStream) -> TokenStream {
+    let input: DeriveInput = syn::parse(input).unwrap();
+
+    let name = input.ident;
+    let proto_struct_name = {
+        input
+            .attrs
+            .iter()
+            .find_map(|attr| {
+                let meta = attr.parse_meta().ok()?;
+                if meta.name() != "protobuf_convert" {
+                    return None;
+                }
+                let list = match meta {
+                    Meta::List(x) => x,
+                    _ => panic!("protobuf_convert attribute expects one argument"),
+                };
+                let name: Path = match list.nested.iter().next().expect("h") {
+                    NestedMeta::Literal(Lit::Str(lit_str)) => lit_str
+                        .parse()
+                        .expect("protobuf_convert argument should be valid type path"),
+                    _ => panic!("protobuf_convert argument should be string"),
+                };
+                Some(name)
+            }).expect("protobuf_convert attribute not set")
+    };
+    let mod_name = Ident::new(&format!("pb_convert_impl_{}", name), Span::call_site());
+    let data = match input.data {
+        Data::Struct(x) => x,
+        _ => panic!("Protobuf convert can be derived for structs only"),
+    };
+
+    let field_names = data
+        .fields
+        .iter()
+        .map(|f| f.ident.clone().unwrap())
+        .collect::<Vec<_>>();
+
+    let to_pb_fn = {
+        let names1 = field_names.clone();
+        let setters = names1
+            .iter()
+            .map(|i| Ident::new(&format!("set_{}", i), Span::call_site()));
+        let names2 = field_names.clone();
+        quote! {
+            fn to_pb(&self) -> Self::ProtoStruct {
+                let mut msg = Self::ProtoStruct::new();
+                #( msg.#setters(self.#names2.to_pb().into()); )*
```

Comment:
If `to_pb()` is not an inherent method (seems it's not), it may make more sense to use its full invocation form specifying the trait.